### PR TITLE
2020 07 14 existing job

### PR DESCRIPTION
**Pull request by tomrittervg**
I think I have all the existing job code working.  Well; that is to say, it passes tests locally. I haven't tried it in-tree yet.

But I think this prep-work branch is good to review.

### CODE REVIEW

File: tests/mock_treeherder_server.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        elif EXPECTEDPATH_FAILURECLASSIFICATION in self.path:
+            self.wfile.write(FAILURE_CLASSIFICATIONS.encode())
+
+        else:
+            prefix = "tests/" if not os.getcwd().endswith("tests") else ""
+            if EXPECTEDPATH_JOBS in self.path:
+                filename = get_appropriate_filename(self.path)
+            elif EXPECTEDPATH_ACTIONSJSON in self.path:
+                filename = "taskcluster_api_response_actionsjson.txt"
```
Comment (metalcanine) [inline tests/mock_treeherder_server.py:81]: We're getting an error in CI due to the file `taskcluster_api_response_actionsjson.txt` missing.

```MockTreeherderServer: Streaming taskcluster_api_response_actionsjson.txt
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 49524)
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.8/3.8.4/Frameworks/Python.framework/Versions/3.8/lib/python3.8/socketserver.py", line 316, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/local/Cellar/python@3.8/3.8.4/Frameworks/Python.framework/Versions/3.8/lib/python3.8/socketserver.py", line 347, in process_request
    self.finish_request(request, client_address)
  File "/usr/local/Cellar/python@3.8/3.8.4/Frameworks/Python.framework/Versions/3.8/lib/python3.8/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/local/Cellar/python@3.8/3.8.4/Frameworks/Python.framework/Versions/3.8/lib/python3.8/socketserver.py", line 720, in __init__
    self.handle()
  File "/usr/local/Cellar/python@3.8/3.8.4/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/server.py", line 427, in handle
    self.handle_one_request()
  File "/usr/local/Cellar/python@3.8/3.8.4/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/server.py", line 415, in handle_one_request
    method()
  File "/Users/jewilde/Source/updatebot/tests/mock_treeherder_server.py", line 87, in do_GET
    with open(prefix + filename, "rb") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'tests/taskcluster_api_response_actionsjson.txt'
----------------------------------------
ERROR```

File: apis/taskcluster.py [hunk 2]
```diff
@@ -52,3 +64,132 @@ def submit_to_try(self, library):
...
+        for j in job_list:
+            if "Gecko Decision Task" == j.job_type_name:
+                decision_task = j
+                break
+        assert decision_task is not None
+
+        # Download its actions.json
+        artifact_url = self.url_taskcluster + "api/queue/v1/task/%s/runs/0/artifacts/public/actions.json" % (decision_task.task_id)
+        r = requests.get(artifact_url, headers=self.HEADERS)
+        actions = r.json()
```
Comment (metalcanine) [inline apis/taskcluster.py:161]: It looks like we might not be getting json back in the response here. Showing up in the testRetrigger test on line 112.

```======================================================================
ERROR: testRetrigger (tests.treeherder_api.TestTaskclusterProvider)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jewilde/Source/updatebot/tests/treeherder_api.py", line 112, in testRetrigger
    decision_task = self.taskclusterProvider.retrigger_jobs(job_list, to_retrigger)
  File "/Users/jewilde/Source/updatebot/components/logging.py", line 31, in func_wrapper
    ret = func(*args, **kwargs)
  File "/Users/jewilde/Source/updatebot/apis/taskcluster.py", line 161, in retrigger_jobs
    actions = r.json()
  File "/Users/jewilde/Library/Caches/pypoetry/virtualenvs/updatebot-6T8aijfh-py3.8/lib/python3.8/site-packages/requests/models.py", line 898, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/local/Cellar/python@3.8/3.8.4/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/local/Cellar/python@3.8/3.8.4/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/Cellar/python@3.8/3.8.4/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)```

File: tests/functionality.py [hunk 3]
```diff
@@ -150,25 +139,47 @@ def testFunctionalityWithRealDatabase(self):
...
+            "./mach vendor": expected_values.library_version_id,
+            "./mach try auto": TRY_OUTPUT(try_revision),
+            "arc diff --verbatim": ARC_OUTPUT
+        }
+        self.configs['Command']['test_mappings'] = command_mappings
+
+        # Make it
+        u = Updatebot(self.configs, self.providers)
+
+        # Ensure we don't have a dirty database with existing jobs
```
Comment (metalcanine) [inline tests/functionality.py:159]: Is there a unittest method we can define that will run before/after each test that we could move database cleanup to?

Something similar to mocha's [beforeEach/afterEach](https://mochajs.org/#asynchronous-hooks)

File: apis/taskcluster.py (hunk 2 above)
Comment (tomrittervg) [reply apis/taskcluster.py:161]: Yeah this is the behavior we get when we get a 404 or other HTTP error.  I'm going to add a commit to handle this type of error all over the place....

Comment (tomrittervg) [reply apis/taskcluster.py]: I didn't add it in a separate commit, instead I added it in-line

File: tests/functionality.py (hunk 3 above)
Comment (tomrittervg) [reply tests/functionality.py:159]: I noodled around on this a bit, wrote a patch, threw it out, wrote a SO question, threw it out, and finally came up with something I think is an improvement.  But it'll be a real bitch to rebase into here, so I'd like to leave it to the end of the next patchset where I add like 4 more tests to this file.

**Comment by tomrittervg** [comment]
Okay I think this will pass. I added a few more tiny patches to the end of the series as well.

**Pull request merged**